packages/medical-domain/src/domains/repositories/medical/medical-data.service.ts

```diff
@@ -355,11 +266,132 @@ class MedicalDataService {
+    }
+    if (data.physicalActivities) {
+      this.medicalData.physicalActivities = this.medicalData.physicalActivities.concat(data.physicalActivities)
+    }
+    if (data.pumpSettings) {
+      this.medicalData.pumpSettings = this.medicalData.pumpSettings.concat(data.pumpSettings)
+    }
+    if (data.reservoirChanges) {
+      this.medicalData.reservoirChanges = this.medicalData.reservoirChanges.concat(data.reservoirChanges)
+    }
+    if (data.smbg) {
+      this.medicalData.smbg = this.medicalData.smbg.concat(data.smbg)
+    }
+    if (data.warmUps) {
+      this.medicalData.warmUps = this.medicalData.warmUps.concat(data.warmUps)
+    }
+    if (data.wizards) {
+      this.medicalData.wizards = this.medicalData.wizards.concat(data.wizards)
+    }
+    if (data.zenModes) {
+      this.medicalData.zenModes = this.medicalData.zenModes.concat(data.zenModes)
+    }
+    if (data.alarmEvents) {
+      this.medicalData.alarmEvents = this.medicalData.alarmEvents.concat(data.alarmEvents)
+    }
+
+    this.deduplicate()
+    this.join()
+    this.setTimeZones()
+    this.group()
+    this.setEndPoints()
+    this.generateFillData()
+    this.generateBasicsData()
+  }
+
+  hasDiabetesData(): boolean {
+    return (
+      this.medicalData.basal.length > 0 ||
+      this.medicalData.bolus.length > 0 ||
+      this.medicalData.cbg.length > 0 ||
+      this.medicalData.smbg.length > 0 ||
+      this.medicalData.wizards.length > 0
+    )
+  }
+
+  filterByDate(epochStart: number, epochEnd: number): Datum[] {
+    return this.data.filter((d) => {
+      const datumEpoch = this.getDatumEpoch(d)
+      return datumEpoch > epochStart && datumEpoch < epochEnd
+    })
+  }
+
+  // Basics data are needed to display the cartridge changes in dashboard and in print actions
+  // We may only need to generate these for the current day displayed in dashboard view and on print...
+  generateBasicsData(startDate: string | null = null, endDate: string | null = null): BasicData | null {
+    const start = startDate ?? this.endpoints[0]
+    const end = endDate ?? this.endpoints[1]
+    let startEpoch = new Date(start).valueOf()
+    let endEpoch = new Date(end).valueOf()
+    if (!endDate) {
+      endEpoch = endEpoch - 1
+    }
+    const endTimezone = this.getTimezoneAt(endEpoch)
+    if (!startDate) {
+      startEpoch = twoWeeksAgo(endEpoch, endTimezone)
+    }
+    const startTimezone = this.getTimezoneAt(startEpoch)
+
+    if (startEpoch > endEpoch) {
+      console.warn('Invalid date range', { start: toISOString(startEpoch), mEnd: toISOString(endEpoch) })
+      this.basicsData = null
+      return null
+    }
+    startEpoch += 1
+    endEpoch -= 1
+
+    this.basicsData = generateBasicData(this.medicalData, startEpoch, startTimezone, endEpoch, endTimezone)
+
+    return this.basicsData
+  }
+
+  addMessage(message: Record<string, unknown>): void {
+    // TODO: cleanup normalize methods
```

Comment:
Is there a JIRA ticket linked to this "TODO"? IMO it is better to have the JIRA ticket and no TODO (you end up finding "TODO" two years later in the code and they are still not done)